File: python-models/src/soma_models/v1/torch/modules/attention.py
import torch
import torch.nn as nn
import torch.nn.functional as F
import math


def apply_rope(
    inputs: torch.Tensor,  # [BATCH_SIZE, SEQ_LEN, NUM_HEADS, HEAD_DIM]
    positions: torch.Tensor,  # [BATCH_SIZE, SEQ_LEN]
    head_dim: int,
    max_wavelength: int = 10_000,
    scale_factor: float = 1.0,
) -> torch.Tensor:
    fraction = 2 * torch.arange(0, head_dim // 2, dtype=inputs.dtype, device=inputs.device) / head_dim
    timescale = max_wavelength ** fraction
    positions = positions.unsqueeze(-1).to(inputs.dtype)  # [batch, seq, 1]
    timescale = timescale.unsqueeze(0).unsqueeze(0)  # [1, 1, head_dim//2]

    sinusoid_inp = positions / timescale
    sinusoid_inp = sinusoid_inp.unsqueeze(-2)  # [batch, seq, 1, head_dim//2]
    if scale_factor < 1.0:
        raise ValueError(f"scale_factor must be >= 1.0, got {scale_factor}")
    sinusoid_inp = sinusoid_inp / scale_factor

    sin = torch.sin(sinusoid_inp)
    cos = torch.cos(sinusoid_inp)

    first_half, second_half = inputs.chunk(2, dim=-1)
    first_part = first_half * cos - second_half * sin
    second_part = second_half * cos + first_half * sin
    out = torch.cat([first_part, second_part], dim=-1)
    return out.to(inputs.dtype)


class MultiHeadAttention(nn.Module):
    def __init__(
        self,
        num_heads: int,
        num_features: int,
        *,
        dropout_rate: float = 0.0,
        use_bias: bool = True,
        max_wavelength: float = 10_000.0,
        scale_factor: float = 1.0,
    ):
        super().__init__()
        self.num_heads = num_heads
        self.num_features = num_features
        self.dropout_rate = dropout_rate
        self.max_wavelength = max_wavelength
        self.scale_factor = scale_factor

        if num_features % num_heads != 0:
            raise ValueError(
                f"Memory dimension ({num_features}) must be divisible by "
                f"'num_heads' heads ({num_heads})."
            )
        self.head_dim = num_features // num_heads

        self.query = nn.Linear(num_features, num_features, bias=use_bias)
        self.key = nn.Linear(num_features, num_features, bias=use_bias)
        self.value = nn.Linear(num_features, num_features, bias=use_bias)
        self.output = nn.Linear(num_features, num_features, bias=use_bias)

    def forward(
        self,
        inputs: torch.Tensor,
        *,
        positions: torch.Tensor | None = None,
    ):
        batch_size, seq_len, _ = inputs.shape

        # Project to Q, K, V and reshape to [batch, seq, heads, head_dim]
        query = self.query(inputs).reshape(batch_size, seq_len, self.num_heads, self.head_dim)
        key = self.key(inputs).reshape(batch_size, seq_len, self.num_heads, self.head_dim)
        value = self.value(inputs).reshape(batch_size, seq_len, self.num_heads, self.head_dim)

        if positions is not None:
            query = apply_rope(query, positions, head_dim=self.head_dim, max_wavelength=self.max_wavelength, scale_factor=self.scale_factor)
            key = apply_rope(key, positions, head_dim=self.head_dim, max_wavelength=self.max_wavelength, scale_factor=self.scale_factor)

        # Transpose to [batch, heads, seq, head_dim] for attention
        query = query.transpose(1, 2)
        key = key.transpose(1, 2)
        value = value.transpose(1, 2)

        # Scale
        scale = math.sqrt(self.head_dim)
        attn_weights = torch.matmul(query, key.transpose(-2, -1)) / scale

        # Generate causal mask: [batch, 1, seq, seq] with True=attend, False=mask
        mask = torch.ones(batch_size, 1, seq_len, seq_len, dtype=torch.bool, device=inputs.device).tril()
        big_neg = torch.finfo(attn_weights.dtype).min
        attn_weights = torch.where(mask, attn_weights, big_neg)

        attn_weights = F.softmax(attn_weights, dim=-1)

        if self.training and self.dropout_rate > 0:
            attn_weights = F.dropout(attn_weights, p=self.dropout_rate)

        # [batch, heads, seq, head_dim]
        attn_output = torch.matmul(attn_weights, value)

        # Transpose back to [batch, seq, heads, head_dim] and reshape
        attn_output = attn_output.transpose(1, 2).reshape(batch_size, seq_len, self.num_features)

        out = self.output(attn_output)
        return out

File: models/src/v1/modules/attention.rs
// Copyright (c) Soma Contributors
// SPDX-License-Identifier: Apache-2.0

use burn::config::Config;
use burn::module::Module;
use burn::nn::attention::generate_autoregressive_mask;
use burn::nn::{Initializer, Linear, LinearConfig};
use burn::prelude::Backend;
use burn::tensor::activation::softmax;
use burn::tensor::{Bool, Int, Tensor};

use crate::v1::{V1_EMBEDDING_DIM, V1_MAX_WAVELENGTH, V1_NUM_HEADS, V1_SCALE_FACTOR};

pub fn apply_rope<B: Backend>(
    inputs: Tensor<B, 4>,         // [batch_size, seq_len, num_heads, head_dim]
    positions: Tensor<B, 2, Int>, // [batch_size, seq_len]
    head_dim: usize,
    max_wavelength: f32,
    scale_factor: f32,
) -> Tensor<B, 4> {
    if scale_factor < 1.0 {
        panic!("scale_factor must be >= 1.0, got {}", scale_factor);
    }
    if !head_dim.is_multiple_of(2) {
        panic!("head_dim must be even, got {}", head_dim);
    }

    // Calculate frequencies: 1 / (max_wavelength ^ (2i / head_dim))
    let device = inputs.device();

    let fraction = Tensor::<B, 1, Int>::arange(0..(head_dim / 2) as i64, &device)
        .float()
        .mul_scalar(2)
        .div_scalar(head_dim as i64);
    let base_tensor = Tensor::full_like(&fraction, max_wavelength);
    let timescale = base_tensor.powf(fraction);

    // Prepare sinusoid input: positions * theta
    let positions_float = positions.float().unsqueeze_dim::<3>(2); // [batch_size, seq_len] -> [batch_size, seq_len, 1]
    let theta = timescale.unsqueeze::<2>().unsqueeze::<3>(); // [head_dim/2] -> [1, head_dim/2] -> [1, 1, head_dim/2]
    let sinusoid_inp = positions_float.div(theta);
    let sinusoid_inp = sinusoid_inp.unsqueeze_dim::<4>(2);
    let sinusoid_inp = sinusoid_inp.div_scalar(scale_factor);
    let sin = sinusoid_inp.clone().sin();
    let cos = sinusoid_inp.cos();

    let chunks = inputs.chunk(2, 3);
    if chunks.len() != 2 {
        panic!("Expected 2 chunks from splitting head_dim, got {}", chunks.len());
    }
    let first_half = chunks[0].clone();
    let second_half = chunks[1].clone();

    let first_part = first_half.clone() * cos.clone() - second_half.clone() * sin.clone();
    let second_part = second_half * cos + first_half * sin;

    Tensor::cat(vec![first_part, second_part], 3)
}

/// Configuration to create a [Multi Head Attention](MultiHeadAttention) layer using the [init function](MultiHeadAttentionConfig::init).
#[derive(Config, Debug)]
pub struct MultiHeadAttentionConfig {
    /// Feature size (same size for input, keys, query, out, etc.)
    #[config(default = "V1_EMBEDDING_DIM")]
    pub num_features: usize,
    /// The number of heads.
    #[config(default = "V1_NUM_HEADS")]
    pub num_heads: usize,
    /// The minimum value a float can take. Default: -1.0e4
    /// This is used to mask attention scores before calculating attention weights.
    /// A value too low might result in NaN.
    #[config(default = -1.0e4)]
    pub min_float: f64,
    /// The type of function used to initialize neural network parameters
    #[config(
        default = "Initializer::KaimingUniform{gain:1.0/num_traits::Float::sqrt(3.0), fan_out_only:false}"
    )]
    pub initializer: Initializer,
    /// The max wavelength for RoPE.
    #[config(default = "V1_MAX_WAVELENGTH")]
    pub max_wavelength: f32,
    /// The RoPE scale factor.
    #[config(default = "V1_SCALE_FACTOR")]
    pub scale_factor: f32,
}

#[derive(Module, Debug)]
pub struct MultiHeadAttention<B: Backend> {
    /// Linear layer to transform the input features into the query space.
    pub query: Linear<B>,
    /// Linear layer to transform the input features into the key space.
    pub key: Linear<B>,
    /// Linear layer to transform the input features into the value space.
    pub value: Linear<B>,
    /// Linear layer to transform the output features back to the original space.
    pub output: Linear<B>,
    /// The size of each linear layer.
    pub num_features: usize,
    /// The number of heads.
    pub num_heads: usize,
    /// The dimension per head.
    pub head_dim: usize,
    /// Minimum value a float can take.
    pub min_float: f64,
    pub max_wavelength: f32,
    pub scale_factor: f32,
}

impl MultiHeadAttentionConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> MultiHeadAttention<B> {
        let linear = |config: &Self| {
            LinearConfig::new(config.num_features, config.num_features)
                .with_initializer(self.initializer.clone())
                .init(device)
        };

        MultiHeadAttention {
            query: linear(self),
            key: linear(self),
            value: linear(self),
            output: linear(self),
            num_features: self.num_features,
            num_heads: self.num_heads,
            head_dim: self.num_features / self.num_heads,
            min_float: self.min_float,
            max_wavelength: self.max_wavelength,
            scale_factor: self.scale_factor,
        }
    }
}

/// [Multihead attention](MultiHeadAttention) forward pass input argument.
#[derive(Debug, Clone)]
pub struct MhaInput<B: Backend> {
    /// Shape `[batch_size, seq_length, num_features]`
    query: Tensor<B, 3>,
    /// Shape `[batch_size, seq_length]`
    positions: Option<Tensor<B, 2, Int>>,
    mask_pad: Option<Tensor<B, 2, Bool>>,
    mask_attn: Option<Tensor<B, 3, Bool>>,
}

impl<B: Backend> MhaInput<B> {
    pub fn new(query: Tensor<B, 3>, positions: Tensor<B, 2, Int>) -> Self {
        let [batch_size, seq_length, _d_model] = query.dims();

        let mask_attn = generate_autoregressive_mask(batch_size, seq_length, &query.device());
        Self { query, positions: Some(positions), mask_attn: Some(mask_attn), mask_pad: None }
    }
}

impl<B: Backend> MultiHeadAttention<B> {
    pub fn forward(&self, input: MhaInput<B>) -> Tensor<B, 3> {
        let [batch_size, seq_length, d_model] = input.query.dims();

        let mut query = self.attention_linear(input.query.clone(), &self.query);
        let mut key = self.attention_linear(input.query.clone(), &self.key);
        let value = self.attention_linear(input.query, &self.value);

        if let Some(positions) = input.positions {
            query = query.swap_dims(1, 2);
            key = key.swap_dims(1, 2);

            query = apply_rope(
                query,
                positions.clone(),
                self.head_dim,
                self.max_wavelength,
                self.scale_factor,
            );
            key = apply_rope(key, positions, self.head_dim, self.max_wavelength, self.scale_factor);

            query = query.swap_dims(1, 2);
            key = key.swap_dims(1, 2);
        }

        let attn_scores = self.attn_scores(query, key);
        let weights = self.attn_weights(attn_scores, input.mask_pad, input.mask_attn);
        let context = weights.clone().matmul(value);
        let context = context.swap_dims(1, 2).reshape([batch_size, seq_length, d_model]);
        self.output.forward(context)
    }

    fn attn_scores(&self, query: Tensor<B, 4>, key: Tensor<B, 4>) -> Tensor<B, 4> {
        query.matmul(key.transpose()).div_scalar((self.head_dim as f32).sqrt())
    }

    fn attn_weights(
        &self,
        mut attn_scores: Tensor<B, 4>,
        mask_pad: Option<Tensor<B, 2, Bool>>,
        mask_attn: Option<Tensor<B, 3, Bool>>,
    ) -> Tensor<B, 4> {
        if let Some(mask_pad) = mask_pad {
            let [batch_size, seq_length] = mask_pad.dims();

            attn_scores = attn_scores
                .mask_fill(mask_pad.reshape([batch_size, 1, 1, seq_length]), self.min_float);
        }

        if let Some(mask_attn) = mask_attn {
            let [batch_size, seq_length_1, seq_length_2] = mask_attn.dims();

            attn_scores = attn_scores.mask_fill(
                mask_attn.reshape([batch_size, 1, seq_length_1, seq_length_2]),
                self.min_float,
            );
        }

        softmax(attn_scores, 3)
    }

    fn attention_linear(&self, x: Tensor<B, 3>, linear: &Linear<B>) -> Tensor<B, 4> {
        let [batch_size, seq_length, _d_model] = x.dims();
        linear
            .forward(x)
            .reshape([batch_size, seq_length, self.num_heads, self.head_dim])
            .swap_dims(1, 2)
    }
}
